crates/matrix-sdk-crypto/src/types/room_history.rs

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2025 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//! Types for sharing encrypted room history, per MSC4268
+
+use std::fmt::Debug;
+
+use ruma::{DeviceKeyAlgorithm, OwnedRoomId};
+use serde::{Deserialize, Serialize};
+use vodozemac::{megolm::ExportedSessionKey, Curve25519PublicKey};
+
+#[cfg(doc)]
+use crate::olm::InboundGroupSession;
+use crate::{
+    olm::ExportedRoomKey,
+    types::{
+        deserialize_curve_key, events::room_key_withheld::RoomKeyWithheldContent,
+        serialize_curve_key, EventEncryptionAlgorithm, SigningKeys,
+    },
+};
+
+/// A bundle of room keys, for sharing encrypted room history.
+#[derive(Deserialize, Serialize, Debug, Default)]
+pub struct RoomKeyBundle {
+    /// Keys that we are sharing with the recipient.
+    pub room_keys: Vec<SharedRoomKey>,
+
+    /// Keys that we are *not* sharing with the recipient.
+    pub withheld: Vec<RoomKeyWithheldContent>,
+}
+
+/// An [`InboundGroupSession`] for sharing as part of the key bundle.
+#[derive(Deserialize, Serialize)]
+pub struct SharedRoomKey {
+    /// The encryption algorithm that the session uses.
+    pub algorithm: EventEncryptionAlgorithm,
+
+    /// The room where the session is used.
+    pub room_id: OwnedRoomId,
+
+    /// The Curve25519 key of the device which initiated the session originally.
+    #[serde(deserialize_with = "deserialize_curve_key", serialize_with = "serialize_curve_key")]
+    pub sender_key: Curve25519PublicKey,
+
+    /// The ID of the session that the key is for.
+    pub session_id: String,
+
+    /// The key for the session.
+    pub session_key: ExportedSessionKey,
+
+    /// The Ed25519 key of the device which initiated the session originally.
+    #[serde(default)]
+    pub sender_claimed_keys: SigningKeys<DeviceKeyAlgorithm>,
+}
+
+impl Debug for SharedRoomKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SharedRoomKey")
+            .field("algorithm", &self.algorithm)
+            .field("room_id", &self.room_id)
+            .field("sender_key", &self.sender_key)
+            .field("session_id", &self.session_id)
+            .field("sender_claimed_keys", &self.sender_claimed_keys)
+            .finish_non_exhaustive()
+    }
+}
```

Comment:
Might make sense to add a snapshot test for this using [`assert_debug_snapshot!()`](https://docs.rs/insta/latest/insta/macro.assert_debug_snapshot.html).

Reply:
done in 36f74ed